### PR TITLE
ShutdownProtobufLibrary when unload paddle frontend dynmaic library t…

### PR DESCRIPTION
**Pull request by meiyang-intel**
…o fix probuf memory leak

### Details:
 - *ShutdownProtobufLibrary when unload frontend dynmaic library which use protobuf to fix memory leak*

### Tickets:
 - *CVS-73232*

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
Can we implement a generic solution working with all protobuf based front ends?
We also have tf and also onnx

**Comment by meiyang-intel** [comment]
Hi @ilya-lavrenov . Have updated the generic solution for the frontends which use protobuf

File: src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    case DLL_THREAD_DETACH:
+        break;
+
+    case DLL_PROCESS_DETACH:
+        google::protobuf::ShutdownProtobufLibrary();
+        break;
+    }
+    return TRUE;  // Successful DLL_PROCESS_ATTACH.
+}
+#elif defined(__linux__)
```
Comment (ilyachur) [inline src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp:22]: Do we need to have something like this for macOS?

File: cmake/developer_package/frontends/frontends.cmake [hunk 2]
```diff
@@ -141,8 +141,13 @@ macro(ov_add_frontend)
     # Disable all warnings for generated code
     set_source_files_properties(${PROTO_SRCS} ${PROTO_HDRS} PROPERTIES COMPILE_OPTIONS -w GENERATED TRUE)
 
+    # Shutdown protobuf when unloading the front dynamic library
+    if(OV_FRONTEND_SHUTDOWN_PROTOBUF AND BUILD_SHARED_LIBS)
```
Comment (ilyachur) [inline cmake/developer_package/frontends/frontends.cmake:145]: I don't see the definition of `OV_FRONTEND_SHUTDOWN_PROTOBUF` in this PR

Comment (meiyang-intel) [reply cmake/developer_package/frontends/frontends.cmake:145]: It's the option defined in #Line 92

File: src/frontends/tensorflow/CMakeLists.txt [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 # TODO: Add LINKABLE_FRONTEND option when tensorflow frontend directory is moved to openvino folder
 ov_add_frontend(NAME tensorflow
+                SHUTDOWN_PROTOBUF
```
Comment (ilya-lavrenov) [inline src/frontends/tensorflow/CMakeLists.txt:7]: do we need this as an option or we can detect this automatically in `ov_add_frontend`? We know, that we have proto files in current directly and can take this into account

File: cmake/developer_package/frontends/frontends.cmake [hunk 4]
```diff
@@ -141,8 +141,13 @@ macro(ov_add_frontend)
...
+        set(SHUTDOWN_PROTOBUF ${PROJECT_SOURCE_DIR}/src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/frontends/frontends.cmake:146]: can we create `INTERFACE` library in cmake and add these sources as `PUBLIC` ones (via https://cmake.org/cmake/help/latest/command/target_sources.html)? 

In this case we can use:
```cmake
target_link_libraries(${TARGET_NAME} PRIVATE ov_protobuf_shutdown)
```
`CMakeLists.txt` with such code can be in `https://github.com/meiyang-intel/openvino/tree/meiyang/fix_protobuf_memory_leak/src/frontends/common/shutdown_protobuf/CMakeLists.txt`

File: src/frontends/tensorflow/CMakeLists.txt (hunk 3 above)
Comment (meiyang-intel) [reply src/frontends/tensorflow/CMakeLists.txt:7]: We can detect automatically by 'if (proto_files) in cmake/developer_package/frontends/frontends.cmake for tensorflow and paddle. But onnx is different, onnx link the protobuf in src/frontends/onnx/onnx_common/CMakeLists.txt. In order to detect the protobuf in onnx, the better way I came up with is to check whether the TARGET_NAME is 'onnx' or check whether link the library 'onnx_common'. But this behavior make 'onnx' an specific target to to add the shutdown files, if so I think it's better to add an option to let target choose whether to add the shutdown files. @ilya-lavrenov what do you think?

File: cmake/developer_package/frontends/frontends.cmake (hunk 4 above)
Comment (meiyang-intel) [reply cmake/developer_package/frontends/frontends.cmake]: have updated

File: src/frontends/tensorflow/CMakeLists.txt (hunk 3 above)
Comment (ilya-lavrenov) [reply src/frontends/tensorflow/CMakeLists.txt:7]: agree, forgot about ONNX differences

**Review by ilya-lavrenov — APPROVED** [review]

**Review by nosovmik — APPROVED** [review]
Verified that no 'memory leak' reports on Windows anymore after loading of Paddle model

File: src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp (hunk 1 above)
Comment (meiyang-intel) [reply src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp]: @ilyachur have updated the code for macOS

**Comment by azhogov** [comment]
post-commit failed on android with the error below. Please fix ASAP or PR will be reverted.
[2022-01-12T10:18:13.469Z] /home/jenkins/agent/workspace/private-ci/ie/build-linux-android_arm/b/repos/openvino/src/frontends/common/shutdown_protobuf/shutdown_protobuf.cpp:1:10: fatal error: 'google/protobuf/stubs/common.h' file not found
[2022-01-12T10:18:13.469Z] #include <google/protobuf/stubs/common.h>
[2022-01-12T10:18:13.469Z]          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Pull request merged**